### PR TITLE
exec supports context (timeout/cancel)

### PR DESCRIPTION
**Pull request by anfernee**
- Add function CommandContext as os/exec does
- Clean up the code a little bit to make it more readable
- Fix panic when calling Stop before Start

Fixes: https://github.com/kubernetes/kubernetes/issues/5644

### CODE REVIEW

**Comment by anfernee** [comment]
cc @apelisse

File: exec/exec.go [hunk 1]
```diff
@@ -110,52 +125,51 @@ func (cmd *cmdWrapper) SetStderr(out io.Writer) {
...
+	if c.Process == nil {
 		return
 	}
+
 	c.Process.Signal(syscall.SIGTERM)
+
 	time.AfterFunc(10*time.Second, func() {
-		if c.ProcessState.Exited() {
-			return
+		if c.ProcessState != nil && !c.ProcessState.Exited() {
```
Comment (thockin) [inline exec/exec.go]: How could this be nil if c.Process was not?  If possible, it deserves a comment

File: exec/exec.go [hunk 2]
```diff
@@ -79,16 +87,23 @@ func New() Interface {
...
 }
 
+// CommandContext is part of the Interface interface.
+func (executor *executor) CommandContext(ctx context.Context, cmd string, args ...string) Cmd {
+	return &cmdWrapper{osexec.CommandContext(ctx, cmd, args...)}
+}
+
 // Wraps exec.Cmd so we can capture errors.
-type cmdWrapper osexec.Cmd
+type cmdWrapper struct {
```
Comment (thockin) [inline exec/exec.go]: Why? This was defined explicitly NOT to inherit methods, and to force us to declare methods we want to pass-through.

File: exec/exec.go [hunk 3]
```diff
@@ -165,10 +179,10 @@ type ExitErrorWrapper struct {
 	*osexec.ExitError
 }
 
-var _ ExitError = ExitErrorWrapper{}
+var _ ExitError = &ExitErrorWrapper{}
 
 // ExitStatus is part of the ExitError interface.
-func (eew ExitErrorWrapper) ExitStatus() int {
+func (eew *ExitErrorWrapper) ExitStatus() int {
```
Comment (apelisse) [inline exec/exec.go]: why is this needed?

File: exec/exec.go [hunk 4]
```diff
@@ -33,6 +34,13 @@ type Interface interface {
 	// This follows the pattern of package os/exec.
 	Command(cmd string, args ...string) Cmd
 
+	// CommandContext returns a Cmd instance which can be used to run a single command.
```
Comment (apelisse) [inline exec/exec.go]: Thank you for documenting this!

**Comment by apelisse** [comment]
`Context` didn't exist in go 1.6. I think we can remove that though. Would you mind opening another PR for that? Thanks.

**Comment by anfernee** [comment]
thanks guys for quick review. I will break it into different commits. also I will reply your concerns here.

File: exec/exec.go (hunk 2 above)
Comment (anfernee) [reply exec/exec.go]: because I feel this casting `(*osexec.Cmd)(cmd).CombinedOutput()` looks weird. if it's defined as embeded type then the casting can be avoided. it's basically the same as `ExitErrorWrapper` below. anyway, I don't have strong opinion on this. it could be as is.

File: exec/exec.go (hunk 1 above)
Comment (anfernee) [reply exec/exec.go]: just to be cautious. there is no documented connection between `c.Process` and `c.ProcessState`. But docs do say it is available after calling Wait or Run. It should be safe. I am okay to leave it as is.

File: exec/exec.go (hunk 3 above)
Comment (anfernee) [reply exec/exec.go]: because the pointer `*ExitErrorWrapper` is the `ExitError`. It passes by reference. previous code also works, but it de-references the pointer and passes into the function by value. Again, not a must-fix. L182 is required though.

**Comment by anfernee** [comment]
close this one, splitted to #27 #28  other improvements are pending for decision.

**Comment by thockin** [comment]
Regarding the cast - it was done that way to avoid inheriting methods.  It
could be done as a non-embedded, but I don't see a point.
